File: _17_letter_combinations_of_a_phone_number/main.py
class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        mapping = {
            '2': ['a', 'b', 'c'],
            '3': ['d', 'e', 'f'],
            '4': ['g', 'h', 'i'],
            '5': ['j', 'k', 'l'],
            '6': ['m', 'n', 'o'],
            '7': ['p', 'q', 'r', 's'],
            '8': ['t', 'u', 'v'],
            '9': ['w', 'x', 'y', 'z']
        }

        if len(digits) == 0:
            return []

        result = []

        def backtracking(i, current):
            if i >= len(digits):
                result.append(''.join(current))
                return

            for letter in mapping[digits[i]]:
                current.append(letter)
                backtracking(i + 1, current)
                current.pop()

        backtracking(0, [])

        return result

File: _17_letter_combinations_of_a_phone_number/main.cpp
class Solution {
public:
    std::vector<string> letterCombinations(std::string digits) {
        std::vector<string> solutions;

        if (digits.size() > 0)
        {
            backtracking(0, digits, solutions, "");
        }
    
        return solutions;
    }

private:
    void backtracking(int i, std::string& digits, std::vector<string>& solutions, std::string current)
    {
        if (i == digits.size())
        {
            solutions.push_back(current);
            return;
        }

        char digit = digits[i];

        for (const char& c : m_Mapping.at(digit))
        {
            current.push_back(c);
            backtracking(i + 1, digits, solutions, current);
            current.pop_back();
        }
    }

    const std::unordered_map<char, std::string> m_Mapping = {
        {'2', "abc"},
        {'3', "def"},
        {'4', "ghi"},
        {'5', "jkl"},
        {'6', "mno"},
        {'7', "pqrs"},
        {'8', "tuv"},
        {'9', "wxyz"}
    };
};